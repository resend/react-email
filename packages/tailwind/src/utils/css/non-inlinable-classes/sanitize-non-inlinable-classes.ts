import { AtRule, type Root, type Rule } from 'postcss';
import selectorParser from 'postcss-selector-parser';
import { sanitizeClassName } from '../../compatibility/sanitize-class-name';

/**
 * This function goes through a few steps to ensure the best email client support and
 * to ensure that the media queries and pseudo classes are going to applied correctly alongisde
 * the inline styles.
 *
 * What it does is:
 * 1. Converts all declarations in all rules into being important ones
 * 2. Sanitizes all the selectors of all rules in the media queries
 * 3. Sanitizes all the selectors of all rules using pseudo classes
 * 4. Merges at rules that have equivalent parameters
 */
export const sanitizeNonInlinableClasses = (root: Root) => {
  const sanitizedRules: (Rule | AtRule)[] = [];
  const nonInlinableClasses: string[] = [];

  // Process rules within at-rules (like media queries)
  root.walkAtRules((atRule) => {
    const sanitizedAtRule = atRule.clone();
    const processedRules: Rule[] = [];

    sanitizedAtRule.walkRules((rule) => {
      const processedSelector = selectorParser((selectorRoot) => {
        sanitizeClasses(selectorRoot, nonInlinableClasses);
      }).processSync(rule.selector);

      const processedRule = rule.clone({ selector: processedSelector });
      processedRule.walkDecls((decl) => {
        decl.important = true;
      });

      processedRules.push(processedRule);
      sanitizedAtRule.removeChild(rule);
    });

    sanitizedAtRule.append(...processedRules);

    const equivalentRule = sanitizedRules.find(
      (r) => r instanceof AtRule && r.params === sanitizedAtRule.params,
    );

    if (equivalentRule) {
      equivalentRule.append(sanitizedAtRule.nodes);
    } else {
      sanitizedRules.push(sanitizedAtRule);
    }
  });

  // Process top-level rules
  root.walkRules((rule) => {
    if (rule.parent && rule.parent.type !== 'root') return;

    let hasPseudoSelector = false as boolean;

    const processedSelector = selectorParser((selectorRoot) => {
      selectorRoot.walkPseudos(() => {
        hasPseudoSelector = true;
      });

      if (!hasPseudoSelector) return;

      sanitizeClasses(selectorRoot, nonInlinableClasses);
    }).processSync(rule.selector);

    if (hasPseudoSelector) {
      const processedRule = rule.clone({ selector: processedSelector });
      processedRule.walkDecls((decl) => {
        decl.important = true;
      });

      sanitizedRules.push(processedRule);
    }
  });

  return {
    nonInlinableClasses,
    sanitizedRules,
  };
};

const sanitizeClasses = (selectorRoot: selectorParser.Root, outputClasses: string[]) => {
  selectorRoot.walkClasses((singleClass) => {
    outputClasses.push(singleClass.value);

    singleClass.replaceWith(
      selectorParser.className({
        ...singleClass,
        value: sanitizeClassName(singleClass.value),
      }),
    );
  });
}
