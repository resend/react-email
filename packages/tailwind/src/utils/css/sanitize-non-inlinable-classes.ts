import { AtRule, type Root, type Rule } from 'postcss';
import selectorParser from 'postcss-selector-parser';
import { sanitizeClassName } from '../compatibility/sanitize-class-name';

/**
 * This function goes through a few steps to ensure the best email client support and
 * to ensure that the media queries and pseudo classes are going to applied correctly alongisde
 * the inline styles.
 *
 * What it does is:
 * 1. Converts all declarations in all rules into being important ones
 * 2. Sanitizes all the selectors of all rules in the media queries
 * 3. Sanitizes all the selectors of all rules using pseudo classes
 * 4. Merges at rules that have equivalent parameters
 */
export const sanitizeNonInlinableClasses = (root: Root) => {
  const sanitizedRules: (Rule | AtRule)[] = [];
  const nonInlinableClasses: string[] = [];

  const selectorProcessor = selectorParser();

  // Process rules within at-rules (like media queries)
  root.walkAtRules((atRule) => {
    const sanitizedAtRule = atRule.clone();

    sanitizedAtRule.walkRules((rule) => {
      const selectorRoot = selectorProcessor.astSync(rule.selector);
      selectorRoot.walkClasses((className) => {
        nonInlinableClasses.push(className.value);
        sanitizeSelectorClassName(className);
      });

      const processedRule = rule.clone({ selector: selectorRoot.toString() });
      processedRule.walkDecls((decl) => {
        decl.important = true;
      });

      rule.replaceWith(processedRule);
    });

    const equivalentRule = sanitizedRules.find(
      (r) => r instanceof AtRule && r.params === sanitizedAtRule.params,
    );

    if (equivalentRule) {
      equivalentRule.append(sanitizedAtRule.nodes);
    } else {
      sanitizedRules.push(sanitizedAtRule);
    }
  });

  // Process top-level rules
  root.walkRules((rule) => {
    if (rule.parent && rule.parent.type !== 'root') return;

    const selectorRoot = selectorProcessor.astSync(rule.selector);

    let hasPseudoSelector = false as boolean;
    selectorRoot.walkPseudos(() => {
      hasPseudoSelector = true;
    });

    if (!hasPseudoSelector) return;

    selectorRoot.walkClasses((className) => {
      nonInlinableClasses.push(className.value);
      sanitizeSelectorClassName(className);
    });

    if (hasPseudoSelector) {
      const processedRule = rule.clone({ selector: selectorRoot.toString() });
      processedRule.walkDecls((decl) => {
        decl.important = true;
      });

      sanitizedRules.push(processedRule);
    }
  });

  return {
    nonInlinableClasses,
    sanitizedRules,
  };
};

const sanitizeSelectorClassName = (className: selectorParser.ClassName) => {
  className.replaceWith(
    className.clone({
      value: sanitizeClassName(className.value),
    }),
  );
};
