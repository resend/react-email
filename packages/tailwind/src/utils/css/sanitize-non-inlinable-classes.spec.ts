import postcss, { Root } from 'postcss';
import { sanitizeNonInlinableClasses } from './sanitize-non-inlinable-classes';

test('sanitizeNonInlinableClasses()', async () => {
  const { root } = await postcss()
    .process(
      `
.ring-gray-900\\/5{
  --tw-ring-color: rgb(17 24 39 / 0.05)
}
.\\[mask-image\\:linear-gradient\\(180deg\\2c white\\2c rgba\\(255\\2c 255\\2c 255\\2c 0\\)\\)\\]{
  -webkit-mask-image: linear-gradient(180deg,white,rgba(255,255,255,0));
          mask-image: linear-gradient(180deg,white,rgba(255,255,255,0))
}

.hover\\:text-sky-600:hover{
  --tw-text-opacity: 1;
  color: rgb(2 132 199 / var(--tw-text-opacity))
}

@media (min-width: 640px){
  .sm\\:mx-auto{
    margin-left: auto;
    margin-right: auto
  }

  .sm\\:focus\\:outline-none:focus{
    outline: 2px solid transparent;
    outline-offset: 2px
  }

  .sm\\:max-w-lg{
    max-width: 32rem
  }

  .sm\\:rounded-lg{
    border-radius: 0.5rem
  }
}

@media (min-width: 768px){
  .md\\:hover\\:bg-gray-100:hover{
    --tw-bg-opacity: 1;
    background-color: rgb(243 244 246 / var(--tw-bg-opacity))
  }

  .md\\:px-10{
    padding-left: 2.5rem;
    padding-right: 2.5rem
  }

  .md\\:py-12{
    padding-top: 3rem;
    padding-bottom: 3rem
  }
}

@media (min-width: 1024px){
  .lg\\:focus\\:underline:focus{
    text-decoration-line: underline
  }
}
`,
    )
    .async();

  const { nonInlinableClasses, sanitizedRules } = sanitizeNonInlinableClasses(
    root as Root,
  );
  expect(nonInlinableClasses).toEqual([
    'sm:mx-auto',
    'sm:focus:outline-none',
    'sm:max-w-lg',
    'sm:rounded-lg',
    'md:hover:bg-gray-100',
    'md:px-10',
    'md:py-12',
    'lg:focus:underline',
    'hover:text-sky-600',
  ]);
  expect(new Root({ nodes: sanitizedRules }).toString()).toBe(`

@media (min-width: 640px){
  .sm_mx-auto{
    margin-left: auto !important;
    margin-right: auto !important
  }

  .sm_focus_outline-none:focus{
    outline: 2px solid transparent !important;
    outline-offset: 2px !important
  }

  .sm_max-w-lg{
    max-width: 32rem !important
  }

  .sm_rounded-lg{
    border-radius: 0.5rem !important
  }
}

@media (min-width: 768px){
  .md_hover_bg-gray-100:hover{
    --tw-bg-opacity: 1 !important;
    background-color: rgb(243 244 246 / var(--tw-bg-opacity)) !important
  }

  .md_px-10{
    padding-left: 2.5rem !important;
    padding-right: 2.5rem !important
  }

  .md_py-12{
    padding-top: 3rem !important;
    padding-bottom: 3rem !important
  }
}

@media (min-width: 1024px){
  .lg_focus_underline:focus{
    text-decoration-line: underline !important
  }
}

.hover_text-sky-600:hover{
  --tw-text-opacity: 1 !important;
  color: rgb(2 132 199 / var(--tw-text-opacity)) !important
}`);
});
