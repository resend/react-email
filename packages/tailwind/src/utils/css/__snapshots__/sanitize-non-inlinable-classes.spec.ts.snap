// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sanitizeNonInlinableClasses() > should css nesting in hover pseudo styles 1`] = `".hover_text-sky-600{&:hover{@media (hover:hover){color:oklch(58.8%0.158 241.966)!important}}}.sm_focus_outline-none{@media (width>=40rem){&:focus{outline-style:none!important}}}.md_hover_bg-gray-100{@media (width>=48rem){&:hover{@media (hover:hover){background-color:oklch(96.7%0.003 264.542)!important}}}}.lg_focus_underline{@media (width>=64rem){&:focus{text-decoration-line:underline!important}}}"`;

exports[`sanitizeNonInlinableClasses() > should css nesting in hover pseudo styles 2`] = `
[
  "hover:text-sky-600",
  "sm:focus:outline-none",
  "md:hover:bg-gray-100",
  "lg:focus:underline",
]
`;

exports[`sanitizeNonInlinableClasses() > should handle rules that can be inlined 1`] = `""`;

exports[`sanitizeNonInlinableClasses() > should handle rules that can be inlined 2`] = `[]`;

exports[`sanitizeNonInlinableClasses() > shuold work with basic media query rules 1`] = `".sm_mx-auto{@media (width>=40rem){margin-inline:auto!important}}.sm_max-w-lg{@media (width>=40rem){max-width:32rem!important}}.sm_rounded-lg{@media (width>=40rem){border-radius:0.5rem!important}}.md_px-10{@media (width>=48rem){padding-inline:2.5rem!important}}.md_py-12{@media (width>=48rem){padding-block:3rem!important}}"`;
